django/core/cache/backends/memcached.py

```diff
@@ -77,8 +77,13 @@ def set(self, key, value, timeout=DEFAULT_TIMEOUT, version=None):
             self._cache.delete(key)
 
     def delete(self, key, version=None):
-        key = self.make_key(key, version=version)
-        self._cache.delete(key)
+
+        key_made = self.make_key(key, version=version)
+        key_exists = bool(self._cache.get(self.make_key(key)))
```

Comment:
We want to avoid a `GETS ` here. What's the rationale for not simply returning `self._cache.delete(key)`?

Reply:
The problem here is that Memcache always returns a `1` if the `delete` operation is successfully executed, independent of whether the value was deleted or not. For non-existent values, it returns a `1` as well. Without a check to see if the key exists, the test `tests.cache.tests.test_delete_returns_correct_value` fails on the last line: https://github.com/django/django/pull/11883/files#diff-b5f0419b57a7b4ef80afcbd9a9f60751R326
Though, to be fair, this is a test I added in this Pull Request.

Reply:
> The problem here is that Memcache always returns a 1 if the delete operation is successfully executed, independent of whether the value was deleted or not. For non-existent values, it returns a 1 as well.

[That's not normal per-memcached protocol](https://github.com/memcached/memcached/blob/master/doc/protocol.txt)

```
Deletion
--------

The command "delete" allows for explicit deletion of items:

delete <key> [noreply]\r\n

- <key> is the key of the item the client wishes the server to delete

- "noreply" optional parameter instructs the server to not send the
  reply.  See the note in Storage commands regarding malformed
  requests.

The response line to this command can be one of:

- "DELETED\r\n" to indicate success

- "NOT_FOUND\r\n" to indicate that the item with this key was not
  found.
```

Which memcached client are you using locally?

Reply:
`python-mecached` should return the right value as long as we don't pass `noreply=True` to `delete`

https://github.com/linsomniac/python-memcached/blob/bad41222379102e3f18f6f2f7be3ee608de6fbff/memcache.py#L548-L575

And `pylibmc` should work as well

https://github.com/lericson/pylibmc/blob/d8bafe91f57ebdd73f76eb80bdb4d4a515c7fc48/src/_pylibmcmodule.c#L1445-L1451

Reply:
@charettes Thanks for the links.
I am using python-memcached locally for testing and overwriting: `CACHES` in `django/conf/global_settings.py`:
```
 CACHES = {
     'default': {
-        'BACKEND': 'django.core.cache.backends.locmem.LocMemCache',
+        'BACKEND': 'django.core.cache.backends.memcached.MemcachedCache',
+        'LOCATION': '127.0.0.1:11211',
     }
 }
```
I just tested with pylibmc and django-pymemcached (https://github.com/django-pymemcache/django-pymemcache), and they both work! 

The problem, as I see it is with the behavior of `python-memcached`:  https://github.com/linsomniac/python-memcached/blob/bad41222379102e3f18f6f2f7be3ee608de6fbff/memcache.py#L530

The single logic line sends `[b'DELETED', b'NOT_FOUND']` as `expected` results from memcache for the `delete` command. When memcache doesn't find the key, it returns `b'NOT_FOUND'`.
However, since this is given as an `expected` value, it returns a `1`. The check happens in this line: https://github.com/linsomniac/python-memcached/blob/bad41222379102e3f18f6f2f7be3ee608de6fbff/memcache.py#L567)

Essentially, python-memcached will return a `1` if the key is deleted OR not found.

Reply:
Well that's unfortunate. I see a few options.

- Make the `python-memcached` client call `_deletetouch(expected=[b'DELETED']` directly.
- Report an issue upstream to change the behavior of `delete`
- Report an issue upstream to allow `delete` to accept a kwarg to change what gets passed to `expected`.

Reply:
I reported an issue https://github.com/linsomniac/python-memcached/issues/170, but I also agree that using `_deletetouch()` is fine for now.